File: 2_3.hs
module EvenFibThree where

main :: IO()
main = do
  let n = 4000000
  print $ sum (takeWhile (<=n) [fib x | x <- [2,5..]])

fibTuple :: (Integer, Integer, Integer) -> (Integer, Integer, Integer)
fibTuple (x, y, 0) = (x, y, 0)
fibTuple (x, y, z) = fibTuple (y, x+y, z-1)

fibResult :: (Integer, Integer, Integer) -> Integer
fibResult (x, y, z) = x

fib :: Integer -> Integer
fib x = fibResult (fibTuple (1, 1, x))

File: 2_3.jl
function fibTuple(t::Tuple{BigInt, BigInt, BigInt})::Tuple{BigInt,BigInt,BigInt}
  (x,y,z) = t
  if z == 0
    return t
  else
    return fibTuple((y, x+y, z-1))
  end
end

function fibResult(t::Tuple{BigInt,BigInt,BigInt})::BigInt
  return t[1]
end

function fib(n::BigInt)::BigInt
  a = BigInt(1)
  return fibResult(fibTuple((a,a,n)))
end
